File: src/authentication.ts
import type { RecordActor } from "./actor.ts";
import {
  validateSecurityAccessRecord,
  type AuthenticationBlockReason,
  type AuthenticationFailureReason,
  type AuthenticationFlowKind,
  type AuthenticationMethod,
  type AuthenticationPrincipalState,
  type AuthorizationDenialReason,
  type Correlation,
  type SecurityAccessRecord,
} from "./records.ts";

export type SecurityAccessMetadata = Correlation &
  RecordActor & {
    readonly access_id: string;
    readonly occurred_at: string;
    readonly source_ip: string;
  };

export interface AuthenticationContext {
  readonly flow_kind?: AuthenticationFlowKind;
  readonly authentication_method?: AuthenticationMethod;
  readonly principal_state?: AuthenticationPrincipalState;
  readonly provider?: string;
}

export function buildAuthenticationSucceededRecord(
  metadata: SecurityAccessMetadata,
  authentication: Required<
    Pick<
      AuthenticationContext,
      "flow_kind" | "authentication_method" | "principal_state"
    >
  > &
    Pick<AuthenticationContext, "provider">,
): SecurityAccessRecord {
  return buildSecurityAccessRecord(metadata, {
    action: "authentication.succeeded",
    ...authentication,
  });
}

export function buildAuthenticationFailedRecord(
  metadata: SecurityAccessMetadata,
  authentication: Required<
    Pick<AuthenticationContext, "flow_kind" | "authentication_method">
  > &
    Pick<AuthenticationContext, "provider">,
  reason: AuthenticationFailureReason,
): SecurityAccessRecord {
  return buildSecurityAccessRecord(metadata, {
    action: "authentication.failed",
    ...authentication,
    reason,
  });
}

export function buildAuthenticationBlockedRecord(
  metadata: SecurityAccessMetadata,
  authentication: AuthenticationContext,
  reason: AuthenticationBlockReason,
): SecurityAccessRecord {
  return buildSecurityAccessRecord(metadata, {
    action: "authentication.blocked",
    ...authentication,
    reason,
  });
}

export function buildAuthorizationDeniedRecord(
  metadata: SecurityAccessMetadata,
  procedure: string,
  reason: AuthorizationDenialReason,
): SecurityAccessRecord {
  return buildSecurityAccessRecord(metadata, {
    action: "authorization.denied",
    attempted_action: procedure,
    permission: "procedure:invoke",
    reason,
  });
}

export function buildMemberCollectionAccessedRecord(
  metadata: SecurityAccessMetadata,
): SecurityAccessRecord {
  return buildPersonalDataAccessedRecord(
    metadata,
    "member_collection",
    "1",
    "member_administration",
  );
}

export function buildMemberAccessedRecord(
  metadata: SecurityAccessMetadata,
  memberId: string,
): SecurityAccessRecord {
  return buildPersonalDataAccessedRecord(
    metadata,
    "member",
    memberId,
    "member_administration",
  );
}

export function buildCampaignRecipientsAccessedRecord(
  metadata: SecurityAccessMetadata,
  campaignId: string,
): SecurityAccessRecord {
  return buildPersonalDataAccessedRecord(
    metadata,
    "campaign",
    campaignId,
    "campaign_recipients",
  );
}

export function buildFormSubmissionsAccessedRecord(
  metadata: SecurityAccessMetadata,
  formId: string,
): SecurityAccessRecord {
  return buildPersonalDataAccessedRecord(
    metadata,
    "form",
    formId,
    "form_submissions",
  );
}

export function buildFormSubmissionAccessedRecord(
  metadata: SecurityAccessMetadata,
  submissionId: string,
): SecurityAccessRecord {
  return buildPersonalDataAccessedRecord(
    metadata,
    "form_submission",
    submissionId,
    "form_submission",
  );
}

function buildPersonalDataAccessedRecord(
  metadata: SecurityAccessMetadata,
  subjectType: string,
  subjectId: string,
  dataCategory: string,
): SecurityAccessRecord {
  return buildSecurityAccessRecord(metadata, {
    action: "personal_data.accessed",
    subject_type: subjectType,
    subject_id: subjectId,
    access_kind: "read",
    data_category: dataCategory,
  });
}

export function authenticationMethodFromKratos(
  value: string,
): AuthenticationMethod {
  switch (value) {
    case "code":
      return "email_code";
    case "oidc":
      return "oidc";
    case "passkey":
      return "passkey";
    default:
      throw new TypeError("unsupported authentication method");
  }
}

function buildSecurityAccessRecord(
  metadata: SecurityAccessMetadata,
  attributes: Omit<
    SecurityAccessRecord,
    keyof SecurityAccessMetadata | keyof Correlation | keyof RecordActor
  >,
): SecurityAccessRecord {
  const record = { ...metadata, ...attributes } as SecurityAccessRecord;
  validateSecurityAccessRecord(record);
  return record;
}

File: authentication.go
package telemetry

import (
	"errors"
	"time"
)

var ErrUnsupportedAuthenticationMethod = errors.New("unsupported authentication method")

// SecurityAccessMetadata contains the fields shared by all durable Security
// Access records. The owning service supplies the durable record ID and fact
// time so builders remain deterministic and transaction-friendly.
type SecurityAccessMetadata struct {
	AccessID   string
	OccurredAt time.Time
	Correlation
	RecordActor
	SourceIP string
}

type AuthenticationContext struct {
	FlowKind             AuthenticationFlowKind
	AuthenticationMethod AuthenticationMethod
	PrincipalState       AuthenticationPrincipalState
	Provider             string
}

func NewAuthenticationSucceededRecord(metadata SecurityAccessMetadata, authentication AuthenticationContext) (SecurityAccessRecord, error) {
	return newValidatedSecurityAccessRecord(metadata, SecurityAccessRecord{
		Action:               SecurityAuthenticationSucceeded,
		FlowKind:             authentication.FlowKind,
		AuthenticationMethod: authentication.AuthenticationMethod,
		PrincipalState:       authentication.PrincipalState,
		Provider:             authentication.Provider,
	})
}

func NewAuthenticationFailedRecord(metadata SecurityAccessMetadata, authentication AuthenticationContext, reason AuthenticationFailureReason) (SecurityAccessRecord, error) {
	return newValidatedSecurityAccessRecord(metadata, SecurityAccessRecord{
		Action:               SecurityAuthenticationFailed,
		FlowKind:             authentication.FlowKind,
		AuthenticationMethod: authentication.AuthenticationMethod,
		Provider:             authentication.Provider,
		Reason:               string(reason),
	})
}

func NewAuthenticationBlockedRecord(metadata SecurityAccessMetadata, authentication AuthenticationContext, reason AuthenticationBlockReason) (SecurityAccessRecord, error) {
	return newValidatedSecurityAccessRecord(metadata, SecurityAccessRecord{
		Action:               SecurityAuthenticationBlocked,
		FlowKind:             authentication.FlowKind,
		AuthenticationMethod: authentication.AuthenticationMethod,
		Provider:             authentication.Provider,
		Reason:               string(reason),
	})
}

func NewAuthorizationDeniedRecord(metadata SecurityAccessMetadata, procedure string, reason AuthorizationDenialReason) (SecurityAccessRecord, error) {
	return newValidatedSecurityAccessRecord(metadata, SecurityAccessRecord{
		Action:          SecurityAuthorizationDenied,
		AttemptedAction: procedure,
		Permission:      AuthorizationProcedureInvokePermission,
		Reason:          string(reason),
	})
}

func NewMemberCollectionAccessedRecord(metadata SecurityAccessMetadata) (SecurityAccessRecord, error) {
	return newPersonalDataAccessedRecord(metadata, "member_collection", "1", "member_administration")
}

func NewMemberAccessedRecord(metadata SecurityAccessMetadata, memberID string) (SecurityAccessRecord, error) {
	return newPersonalDataAccessedRecord(metadata, "member", memberID, "member_administration")
}

func NewCampaignRecipientsAccessedRecord(metadata SecurityAccessMetadata, campaignID string) (SecurityAccessRecord, error) {
	return newPersonalDataAccessedRecord(metadata, "campaign", campaignID, "campaign_recipients")
}

func NewFormSubmissionsAccessedRecord(metadata SecurityAccessMetadata, formID string) (SecurityAccessRecord, error) {
	return newPersonalDataAccessedRecord(metadata, "form", formID, "form_submissions")
}

func NewFormSubmissionAccessedRecord(metadata SecurityAccessMetadata, submissionID string) (SecurityAccessRecord, error) {
	return newPersonalDataAccessedRecord(metadata, "form_submission", submissionID, "form_submission")
}

func newPersonalDataAccessedRecord(metadata SecurityAccessMetadata, subjectType, subjectID, dataCategory string) (SecurityAccessRecord, error) {
	return newValidatedSecurityAccessRecord(metadata, SecurityAccessRecord{
		Action:       SecurityPersonalDataAccessed,
		SubjectType:  subjectType,
		SubjectID:    subjectID,
		AccessKind:   PersonalDataAccessRead,
		DataCategory: dataCategory,
	})
}

func AuthenticationMethodFromKratos(value string) (AuthenticationMethod, error) {
	var method AuthenticationMethod
	switch value {
	case "code":
		method = AuthenticationMethodEmailCode
	case "oidc":
		method = AuthenticationMethodOIDC
	case "passkey":
		method = AuthenticationMethodPasskey
	default:
		return "", ErrUnsupportedAuthenticationMethod
	}
	return method, nil
}

func newValidatedSecurityAccessRecord(metadata SecurityAccessMetadata, attributes SecurityAccessRecord) (SecurityAccessRecord, error) {
	attributes.AccessID = metadata.AccessID
	attributes.OccurredAt = metadata.OccurredAt
	attributes.Correlation = metadata.Correlation
	attributes.RecordActor = metadata.RecordActor
	attributes.SourceIP = metadata.SourceIP
	if err := attributes.Validate(); err != nil {
		return SecurityAccessRecord{}, err
	}
	return attributes, nil
}
